File: mobile/lib/features/guides/guide_body.dart
/// Markdown-lite as DATA (5.7's parser ported to Dart): the authored bodies
/// use only headings/paragraphs/bullets, so a full markdown package - and its
/// HTML questions - stays unearned. The widget renders segments; no markup
/// ever reaches the tree.
sealed class GuideBodySegment {
  const GuideBodySegment(this.text);
  final String text;
}

class HeadingSegment extends GuideBodySegment {
  const HeadingSegment(super.text);
}

class ParagraphSegment extends GuideBodySegment {
  const ParagraphSegment(super.text);
}

class BulletSegment extends GuideBodySegment {
  const BulletSegment(super.text);
}

List<GuideBodySegment> parseGuideBody(String markdown) {
  final segments = <GuideBodySegment>[];
  final paragraph = <String>[];

  void flush() {
    if (paragraph.isNotEmpty) {
      segments.add(ParagraphSegment(paragraph.join(' ')));
      paragraph.clear();
    }
  }

  for (final raw in markdown.replaceAll('\r\n', '\n').split('\n')) {
    final line = raw.trim();
    if (line.isEmpty) {
      flush();
    } else if (RegExp(r'^#{1,6}\s').hasMatch(line)) {
      flush();
      segments.add(HeadingSegment(line.replaceFirst(RegExp(r'^#{1,6}\s+'), '')));
    } else if (RegExp(r'^[-*]\s').hasMatch(line)) {
      flush();
      segments.add(BulletSegment(line.replaceFirst(RegExp(r'^[-*]\s+'), '')));
    } else {
      paragraph.add(line); // hard-wrapped source lines join into one paragraph
    }
  }
  flush();

  return segments;
}

File: frontend/libs/guides/src/lib/guide-body.ts
/**
 * Markdown-lite for guide bodies, as DATA not HTML: the authored content uses only headings and
 * paragraphs (see the seeded guide), so a full markdown pipe - and the innerHTML sanitization
 * questions it drags in - is not earned yet. The template renders segments with @for/@switch;
 * nothing ever reaches the DOM as markup.
 */
export interface GuideBodySegment {
  kind: 'heading' | 'paragraph' | 'bullet';
  text: string;
}

export function parseGuideBody(markdown: string): GuideBodySegment[] {
  const segments: GuideBodySegment[] = [];
  let paragraph: string[] = [];

  const flush = () => {
    if (paragraph.length) {
      segments.push({ kind: 'paragraph', text: paragraph.join(' ') });
      paragraph = [];
    }
  };

  for (const raw of markdown.replace(/\r\n/g, '\n').split('\n')) {
    const line = raw.trim();
    if (!line) {
      flush();
    } else if (/^#{1,6}\s/.test(line)) {
      flush();
      segments.push({ kind: 'heading', text: line.replace(/^#{1,6}\s+/, '') });
    } else if (/^[-*]\s/.test(line)) {
      flush();
      segments.push({ kind: 'bullet', text: line.replace(/^[-*]\s+/, '') });
    } else {
      paragraph.push(line);   // hard-wrapped source lines join into one paragraph
    }
  }
  flush();

  return segments;
}
